File: cpp/tests/testLink.cpp
/**
 * @file  test_link.cpp
 * @brief test link
 * @Author: Frank Dellaert and Mandy Xie
 */

#include <CppUnitLite/TestHarness.h>
#include <DhLink.h>
#include <gtsam/base/Testable.h>
#include <gtsam/base/TestableAssertions.h>
#include <gtsam/linear/VectorValues.h>

using namespace std;
using namespace gtsam;
using namespace manipulator;

static const Vector1 ZERO1 = Vector1::Zero();
static const Vector6 ZERO6 = Vector6::Zero();
DhLink dhLink = DhLink(0, 0, 2, 0, 'R', 1, Point3(-1, 0, 0),
                          (Matrix(3, 3) << 0, 0, 0, 0, 1 / 6.0, 0, 0, 0, 1 / 6.0).finished(), -5, 10, 2);

// Test factors for forward dynamics, middle link of stationary RRR example
TEST(Link, forward_factors) {
  // Create stationary state
  double v2 = 0;
  Vector6 twist_2 = ZERO6;
  double torque_2 = 0.0;

  // Create all factors
  Pose3 jTi = Pose3(Rot3(), Point3(-2, 0, 0));
  Pose3 kTj = Pose3(Rot3(), Point3(-2, 0, 0));
  GaussianFactorGraph factors =
      dhLink.forwardFactors(2, jTi, v2, twist_2, torque_2, kTj);
  EXPECT(assert_equal(factors.size(), 3));

  // Create ground truth values
  VectorValues ground_truth;
  ground_truth.insert(a(2), ZERO1);
  ground_truth.insert(T(1), ZERO6);
  ground_truth.insert(T(2), ZERO6);
  ground_truth.insert(F(2), ZERO6);
  ground_truth.insert(F(3), ZERO6);

  // // Assert that error is zero for ground-truth
  EXPECT(assert_equal(factors.error(ground_truth), 0));
}

// Test factors for forward dynamics, middle link of stationary RRR example
TEST(Link, inverse_factors) {
  // Create stationary state
  double v2 = 0;
  Vector6 twist_2 = ZERO6;
  double acceleration_2 = 0.0;

  // Create all factors
  Pose3 jTi = Pose3(Rot3(), Point3(-2, 0, 0));
  Pose3 kTj = Pose3(Rot3(), Point3(-2, 0, 0));
  GaussianFactorGraph factors =
      dhLink.inverseFactors(2, jTi, v2, twist_2, acceleration_2, kTj);
  EXPECT(assert_equal(factors.size(), 3));

  // // Create ground truth values
  // VectorValues ground_truth;
  // ground_truth.insert(t(2), ZERO1);
  // ground_truth.insert(T(1), ZERO6);
  // ground_truth.insert(T(2), ZERO6);
  // ground_truth.insert(F(2), ZERO6);
  // ground_truth.insert(F(3), ZERO6);

  // // // Assert that error is zero for ground-truth
  // EXPECT(assert_equal(factors.error(ground_truth), 0));
}

int main() {
  TestResult tr;
  return TestRegistry::runAllTests(tr);
}

File: python/tests/test_link.py
"""
 * GTDynamics Copyright 2020, Georgia Tech Research Corporation,
 * Atlanta, Georgia 30332-0415
 * All Rights Reserved
 * See LICENSE for the license information
 *
 * @file  test_link.py
 * @brief Test Link class.
 * @author Frank Dellaert, Varun Agrawal, Mandy Xie, Alejandro Escontrela, and Yetong Zhang
"""

# pylint: disable=no-name-in-module, import-error, no-member
import os.path as osp
import unittest

import numpy as np
from gtsam import Point3, Pose3, Rot3
from gtsam.utils.test_case import GtsamTestCase

import gtdynamics as gtd


class TestLink(GtsamTestCase):
    """Tests for the Link class."""
    def setUp(self):
        """Set up the fixtures."""
        # load example robot
        SDF_PATH = osp.join(osp.dirname(osp.realpath(__file__)), "..", "..",
                            "sdfs")
        self.simple_rr = gtd.CreateRobotFromFile(
            osp.join(SDF_PATH, "test", "simple_rr.sdf"), "simple_rr_sdf")

    def test_params_constructor(self):
        """Check the links in the simple RR robot."""

        l0 = self.simple_rr.link("link_0")
        l1 = self.simple_rr.link("link_1")

        # Both link frames are defined in the world frame.
        self.gtsamAssertEquals(l0.wTl(), Pose3())
        self.gtsamAssertEquals(l1.wTl(), Pose3())

        # Verify center of mass defined in the link frame is correct.
        self.gtsamAssertEquals(l0.lTcom(), Pose3(Rot3(), Point3(0, 0, 0.1)))
        self.gtsamAssertEquals(l1.lTcom(), Pose3(Rot3(), Point3(0, 0, 0.5)))

        # Verify center of mass defined in the world frame is correct.
        self.gtsamAssertEquals(l0.wTcom(), Pose3(Rot3(), Point3(0, 0, 0.1)))
        self.gtsamAssertEquals(l1.wTcom(), Pose3(Rot3(), Point3(0, 0, 0.5)))

        # Verify that mass is correct.
        self.assertEqual(l0.mass(), 0.01)
        self.assertEqual(l1.mass(), 0.01)

        # Verify that inertia elements are correct.
        np.testing.assert_allclose(
            l0.inertia(), np.array([[0.05, 0, 0], [0, 0.06, 0], [0, 0, 0.03]]))
        np.testing.assert_allclose(
            l1.inertia(), np.array([[0.05, 0, 0], [0, 0.06, 0], [0, 0, 0.03]]))

    def test_get_joints(self):
        """Test the joints method."""
        l0 = self.simple_rr.link("link_0")
        l1 = self.simple_rr.link("link_1")

        self.assertIsInstance(l0.joints(), list)
        self.assertIsInstance(l0.joints()[0], gtd.Joint)

        self.assertEqual(len(l0.joints()), 1)
        self.assertEqual(len(l1.joints()), 2)


if __name__ == "__main__":
    unittest.main()
